src/compiler/binder.ts

```diff
@@ -688,10 +688,49 @@ namespace ts {
                     break;
                 default:
                     bindEachChild(node);
+                    if (isFunctionLikeDeclaration(node)) {
+                        bindJSDocParameters(node);
+                    }
                     break;
             }
         }
 
+        function bindJSDocParameters(node: SignatureDeclaration): void {
+            const { parameters } = node;
+            if (parameters.length === 0) {
+                return;
+            }
+
+            // Clear any previous binding
+            for (const p of parameters) {
+                p.jsdocParamTags = undefined;
+            }
+
+            let paramIndex = 0;
+            for (const tag of getJSDocTags(parent)) {

```

Comment:
This is going to be O(mn) where m = |tags| and n = |params|. That's because, for each tag, we examine every parameter starting at 0. Is this necessary? I think it would suffice to increment paramIndex until a single matching tag is found. 

Even better, you could iterate over the parameters instead and keep a non-restarting index into the tags. You can give up when the tags run out.

```hs
zip :: [Tag] -> [Param] -> [Param]
zip (t:tags) (p:params) | name t == name p = p { tags = Just t } : zip tags params
zip (t:tags) params = zip tags params
zip [] params = params
zip tags [] = []
```

Note that the Haskell version doesn't side-effect, it just assumes there is a field `tags :: Maybe Tag` that is set to `Nothing` when this function is called.

```ts
let tagIndex = 0;
for (const param of parameters) {
  while (tags[tagIndex].name.escapedText !== param.name.escapedText && tagIndex < tags.length) {
    tagIndex++;
  }
  if (tagIndex >= tags.length) {
    break;
  }
  if (tags[tagIndex].name.escapedText === param.name.escapedText) {
    param.jsdocParamTags = append(param.jsdocParamTags, tags[tagIndex]);
  }
}
```

Reply:
We want to keep supporting out-of-order `@param` tags, but in the case where they *are* in order this will be `O(n)` because we start looking at the previous `paramIndex` so usually we only go through the loop once.

Reply:
I see. I misunderstood the `const startParamIndex = paramIndex`. This is better than what I proposed, then.